src/transaction_history.rs

```diff
@@ -0,0 +1,500 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use cosmwasm_std::{
+    Api, CanonicalAddr, Coin, HumanAddr, ReadonlyStorage, StdError, StdResult, Storage, Uint128,
+};
+use cosmwasm_storage::{PrefixedStorage, ReadonlyPrefixedStorage};
+
+use secret_toolkit::storage::{AppendStore, AppendStoreMut};
+
+use crate::state::Config;
+
+const PREFIX_TXS: &[u8] = b"transactions";
+const PREFIX_TRANSFERS: &[u8] = b"transfers";
+
+// Note that id is a globally incrementing counter.
+// Since it's 64 bits long, even at 50 tx/s it would take
+// over 11 billion years for it to rollback. I'm pretty sure
+// we'll have bigger issues by then.
+#[derive(Serialize, Deserialize, JsonSchema, Clone, Debug)]
+pub struct Tx {
+    pub id: u64,
+    pub from: HumanAddr,
+    pub sender: HumanAddr,
+    pub receiver: HumanAddr,
+    pub coins: Coin,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub memo: Option<String>,
+    // The block time and block height are optional so that the JSON schema
+    // reflects that some SNIP-20 contracts may not include this info.
+    pub block_time: Option<u64>,
```

Comment:
Shouldn't `block_time` and `block_height` have the `#[serde(skip_serializing_if = "Option::is_none")]` annotation as well?

Reply:
There just wasn't a need since they're always `Some()` in this implementation. I put them down as an `Option` to make the schema reflect they might not be available in other SNIP-20 contracts

Reply:
Yup makes sense, I was thinking about backwards compatibility but thinking of it now it's fine 👍🏼